app/src/main/cpp/skyline/services/codec/IHardwareOpusDecoder.cpp

```diff
@@ -0,0 +1,126 @@
+// SPDX-License-Identifier: MPL-2.0
+// Copyright © 2021 Skyline Team and Contributors (https://github.com/skyline-emu/)
+
+#include "IHardwareOpusDecoder.h"
+#include "results.h"
+
+namespace skyline::service::codec {
+
+    size_t CalculateOutBufferSize(i32 sampleRate, i32 channelCount, i32 frameSize) {
+        return util::AlignUp(frameSize * channelCount / (OpusFullbandSampleRate / sampleRate), 0x40);
+    }
+
+    IHardwareOpusDecoder::IHardwareOpusDecoder(const DeviceState &state, ServiceManager &manager, i32 sampleRate, i32 channelCount) : BaseService(state, manager), sampleRate(sampleRate), channelCount(channelCount), decoderOutputBufferSize(CalculateOutBufferSize(sampleRate, channelCount, MaxFrameSizeNormal)) {
+        int result;
+        decoderState = opus_decoder_create(sampleRate, channelCount, &result);
+        if (result != OPUS_OK)
+            throw OpusException(result);
+    }
+
+    IHardwareOpusDecoder::~IHardwareOpusDecoder() {
+        opus_decoder_destroy(decoderState);
+    }
+
+    Result IHardwareOpusDecoder::DecodeInterleavedOld(type::KSession &session, ipc::IpcRequest &request, ipc::IpcResponse &response) {
+        auto opusDataIn{request.inputBuf.at(0)};
+        auto pcmDataOut{request.outputBuf.at(0).cast<opus_int16>()};
+
+        auto decodeResult{Decode(opusDataIn, pcmDataOut)};
+        if (!decodeResult) {
+            throw OpusException(decodeResult.result);
+        }
+
+        response.Push(decodeResult->decodedDataSize);
+        response.Push(decodeResult->decodedSampleCount);
+        return {};
+    }
+
+    Result IHardwareOpusDecoder::DecodeInterleavedWithPerfOld(type::KSession &session, ipc::IpcRequest &request, ipc::IpcResponse &response) {
+        auto opusDataIn{request.inputBuf.at(0)};
+        auto pcmDataOut{request.outputBuf.at(0).cast<opus_int16>()};
+
+        auto decodeResult{Decode(opusDataIn, pcmDataOut)};
+        if (!decodeResult) {
+            throw OpusException(decodeResult.result);
+        }
+
+        response.Push(decodeResult->decodedDataSize);
+        response.Push(decodeResult->decodedSampleCount);
+        response.Push<u64>(0);
```

Comment:
this is a microsecond timer for how long the decode took, you can use chrono for that